### PR TITLE
changes build_image.sh to push to gcr

### PR DESCRIPTION
**Pull request by swiftdiaries**
Fixes https://github.com/kubeflow/kubeflow/issues/1575, where building centraldashboard image remains stuck forever.

Related to #1435 Update Central UI image

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kubeflow/1577)

### CODE REVIEW

File: components/centraldashboard/build_image.sh [hunk 1]
```diff
@@ -18,6 +21,9 @@ done
 docker build -f ${DOCKERFILE} -t ${IMAGE}:${TAG} .
 
 gcloud auth activate-service-account --key-file=${GOOGLE_APPLICATION_CREDENTIALS}
-gcloud docker -- push "${IMAGE}:${TAG}"
-docker tag "${IMAGE}:${TAG}" "${IMAGE}:latest"
-gcloud docker -- push "${IMAGE}:latest"
\ No newline at end of file
+
```
Comment (jlewi) [inline components/centraldashboard/build_image.sh]: Don't you want to replace docker build with use of GCB? It seems like the problem is the docker build isn't working so we need to switch to using GCB to build.

Here are some instructions
https://cloud.google.com/cloud-build/docs/running-builds/start-build-manually
Should be something like
gcloud builds submit --tag gcr.io/[PROJECT_ID]/[IMAGE_NAME] .

**Comment by jlewi** [comment]
Can you also enable the image release workflow in the presubmit so we can verify its passin?
Uncomment this line:
https://github.com/kubeflow/kubeflow/blob/master/prow_config.yaml#L65

File: components/centraldashboard/build_image.sh (hunk 1 above)
Comment (swiftdiaries) [reply components/centraldashboard/build_image.sh]: The previous build_image.sh didn't have a project ID to push to and was failing as a result.

**Comment by jlewi** [comment]
The build step still isn't generating any output and appears to be stuck. That's very weird.
It looks like the code is still calling docker build; so if it is a problem with docker in docker this won't fix it.
I'd suggest adding an echo statement before & after docker build just to confirm its getting invoked.

**Comment by jlewi** [comment]
Looks like the pods is stuck in the pending state because of insufficient resources.

```
kubectl describe pods kubeflow-presubmit-dashboard-release-1577-a5969ad-3426-c12a-1271308836
Name:           kubeflow-presubmit-dashboard-release-1577-a5969ad-3426-c12a-1271308836
Namespace:      kubeflow-test-infra
Node:           <none>
Labels:         workflows.argoproj.io/completed=false
                workflows.argoproj.io/workflow=kubeflow-presubmit-dashboard-release-1577-a5969ad-3426-c12a
Annotations:    workflows.argoproj.io/node-name=kubeflow-presubmit-dashboard-release-1577-a5969ad-3426-c12a.image-build-release
                workflows.argoproj.io/template={"name":"image-build-release","inputs":{},"outputs":{},"metadata":{},"container":{"name":"","image":"gcr.io/kubeflow-ci/test-worker:latest","command":["/bin/bash","-c","...
Status:         Pending
IP:             
Controlled By:  Workflow/kubeflow-presubmit-dashboard-release-1577-a5969ad-3426-c12a
Containers:
  main:
    Image:  gcr.io/kubeflow-ci/test-worker:latest
    Port:   <none>
    Command:
      /bin/bash
      -c
      /mnt/test-data-volume/kubeflow-presubmit-dashboard-release-1577-a5969ad-3426-c12a/src/kubeflow/kubeflow/components/centraldashboard/build_image.sh /mnt/test-data-volume/kubeflow-presubmit-dashboard-release-1577-a5969ad-3426-c12a/src/kubeflow/kubeflow/components/centraldashboard/Dockerfile gcr.io/kubeflow-ci/centraldashboard latest
    Limits:
      cpu:     16
      memory:  32Gi
    Requests:
      cpu:     2
      memory:  2Gi
    Environment:
      PYTHONPATH:                      /mnt/test-data-volume/kubeflow-presubmit-dashboard-release-1577-a5969ad-3426-c12a/src/kubeflow/kubeflow:/mnt/test-data-volume/kubeflow-presubmit-dashboard-release-1577-a5969ad-3426-c12a/src/kubeflow/testing/py
      GOOGLE_APPLICATION_CREDENTIALS:  /secret/gcp-credentials/key.json
      GITHUB_TOKEN:                    <set to the key 'github_token' in secret 'github-token'>  Optional: false
      BUILD_ID:                        3426
      BUILD_NUMBER:                    3426
      JOB_NAME:                        kubeflow-presubmit
      JOB_TYPE:                        presubmit
      PULL_BASE_SHA:                   61c7da4756c8ba149637bf4a9e1805258fbead96
      PULL_NUMBER:                     1577
      PULL_PULL_SHA:                   a5969ad5230c0973d878d9ac01c356dc427fca47
      REPO_NAME:                       kubeflow
      REPO_OWNER:                      kubeflow
      DOCKER_HOST:                     127.0.0.1
    Mounts:
      /mnt/test-data-volume from kubeflow-test-volume (rw)
      /secret/gcp-credentials from gcp-credentials (rw)
      /secret/github-token from github-token (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-tdkq7 (ro)
  wait:
    Image:  argoproj/argoexec:v2.1.0
    Port:   <none>
    Command:
      argoexec
    Args:
      wait
    Environment:
      ARGO_POD_IP:      (v1:status.podIP)
      ARGO_POD_NAME:   kubeflow-presubmit-dashboard-release-1577-a5969ad-3426-c12a-1271308836 (v1:metadata.name)
      ARGO_NAMESPACE:  kubeflow-test-infra (v1:metadata.namespace)
    Mounts:
      /argo/podmetadata from podmetadata (rw)
      /var/lib/docker from docker-lib (ro)
      /var/run/docker.sock from docker-sock (ro)
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-tdkq7 (ro)
  dind:
    Image:  docker:17.10-dind
    Port:   <none>
    Limits:
      cpu:     16
      memory:  32Gi
    Requests:
      cpu:        2
      memory:     2Gi
    Environment:  <none>
    Mounts:
      /mnt/test-data-volume from kubeflow-test-volume (rw)
      /secret/gcp-credentials from gcp-credentials (rw)
      /secret/github-token from github-token (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-tdkq7 (ro)
Conditions:
  Type           Status
  PodScheduled   False 
Volumes:
  podmetadata:
    Type:  DownwardAPI (a volume populated by information about the pod)
    Items:
      metadata.annotations -> annotations
  docker-lib:
    Type:          HostPath (bare host directory volume)
    Path:          /var/lib/docker
    HostPathType:  Directory
  docker-sock:
    Type:          HostPath (bare host directory volume)
    Path:          /var/run/docker.sock
    HostPathType:  Socket
  kubeflow-test-volume:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  nfs-external
    ReadOnly:   false
  github-token:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  github-token
    Optional:    false
  gcp-credentials:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  gcp-credentials
    Optional:    false
  default-token-tdkq7:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  default-token-tdkq7
    Optional:    false
QoS Class:       Burstable
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:
  Type     Reason             Age                 From                Message
  ----     ------             ----                ----                -------
  Warning  FailedScheduling   3m (x1328 over 6h)  default-scheduler   0/6 nodes are available: 3 Insufficient cpu, 3 PodToleratesNodeTaints.
  Normal   NotTriggerScaleUp  3m (x2192 over 6h)  cluster-autoscaler  pod didn't trigger scale-up (it wouldn't fit if a new node is added)
```

**Comment by jlewi** [comment]
The nodes have 8 CPUs so I'm guessing its the limit value of 16 that is causing the problem. I think that's my fault. I bumped it to 16 CPUs to see if that would fix the problem. I'd suggest decreasing it to 8 and seeing if it gets scheduled.

**Comment by jlewi** [comment]
Also the script shouldn't hardcode the project/registry it pushes to. This should be a command line argument that is passed through a workflow parameter.
This is because in presubmits we want to push to kubeflow-ci and in postsubmits we want to push to gcr.io/kubeflow-images-public

Registry should be specified like this
https://github.com/kubeflow/kubeflow/blob/61c7da4756c8ba149637bf4a9e1805258fbead96/prow_config.yaml#L95

i.e. the parameter should be "gcr.io/kubeflow-ci" and not the project name.

**Comment by swiftdiaries** [comment]
My initial commit was without the project repo hard coded in then I assumed
wrongly that that was breaking the script.
I’ll change that back.
Thank you for the review!

**Comment by jlewi** [comment]
Thanks. Let me know when this is ready for another look.

**Comment by richardsliu** [comment]
FYI the presubmit failure is due to https://github.com/kubeflow/testing/issues/209.

**Comment by jlewi** [comment]
kubeflow/testing#209 is one issue
but it looks like the pods aren't being scheduled so the build isn't actually happening.

See previous comment:
https://github.com/kubeflow/kubeflow/pull/1577#issuecomment-423397313

The pod isn't being scheduled because of insufficient resources.

```
Events:
  Type     Reason             Age                 From                Message
  ----     ------             ----                ----                -------
  Warning  FailedScheduling   33m (x9 over 35m)   default-scheduler   0/6 nodes are available: 1 NodeUnderDiskPressure, 3 Insufficient cpu, 3 PodToleratesNodeTaints.
  Warning  FailedScheduling   28m (x15 over 31m)  default-scheduler   0/6 nodes are available: 2 NodeUnderDiskPressure, 3 Insufficient cpu, 3 PodToleratesNodeTaints.
  Warning  FailedScheduling   4m (x108 over 44m)  default-scheduler   0/6 nodes are available: 3 Insufficient cpu, 3 PodToleratesNodeTaints.
  Normal   NotTriggerScaleUp  3m (x218 over 44m)  cluster-autoscaler  pod didn't trigger scale-up (it wouldn't fit if a new node is added)
```

**Comment by jlewi** [comment]
Looks like the pod got scheduled and ran but failed with

```
/mnt/test-data-volume/kubeflow-presubmit-dashboard-release-1577-5c5295b-3449-19a8/src/kubeflow/kubeflow/components/centraldashboard/build_image.sh: line 1
4: GCP_REGISTRY: unbound variable
```

**Comment by swiftdiaries** [comment]
Closing because of the git mess.

**Pull request closed without merging**